File: frontend/src/crdt/PNCounter.ts
/**
 * PN-Counter (Positive-Negative Counter) — TypeScript-Version.
 *
 * Identische Logik wie PNCounter.java im Backend.
 * Verwendet für: Kontaktaufnahmen-Zähler pro Kunde.
 *
 * Jeder Node trackt seine eigenen Increment/Decrement-Zähler separat.
 * Merge: Pro Node das Maximum nehmen (nicht die Summe!).
 * Wert = sum(increments) - sum(decrements)
 */
export class PNCounter {
  // Pro Node: wie oft hat DIESER Node incrementiert
  private incs = new Map<string, number>();
  // Pro Node: wie oft hat DIESER Node decrementiert
  private decs = new Map<string, number>();

  /** Increment um 1 für den gegebenen Node */
  increment(nodeId: string): void {
    this.incs.set(nodeId, (this.incs.get(nodeId) ?? 0) + 1);
  }

  /** Decrement um 1 für den gegebenen Node */
  decrement(nodeId: string): void {
    this.decs.set(nodeId, (this.decs.get(nodeId) ?? 0) + 1);
  }

  /** Aktueller Zählerwert: Summe aller Increments minus Summe aller Decrements */
  get value(): number {
    let inc = 0;
    for (const v of this.incs.values()) inc += v;
    let dec = 0;
    for (const v of this.decs.values()) dec += v;
    return inc - dec;
  }

  /**
   * Merged einen anderen Counter: Pro Node das Maximum nehmen.
   * Max statt Summe, weil ein Node seinen Zähler nur erhöht —
   * bei max(3,2)=3 ist klar dass der Node 3x incrementiert hat, nicht 3+2=5.
   */
  merge(other: PNCounter): void {
    for (const [nodeId, count] of other.incs) {
      this.incs.set(nodeId, Math.max(this.incs.get(nodeId) ?? 0, count));
    }
    for (const [nodeId, count] of other.decs) {
      this.decs.set(nodeId, Math.max(this.decs.get(nodeId) ?? 0, count));
    }
  }
}

File: backend/src/main/java/com/collabcrm/crdt/PNCounter.java
package com.collabcrm.crdt;

import java.util.HashMap;
import java.util.Map;

/**
 * PN-Counter (Positive-Negative Counter) — CRDT für Zähler.
 */
public class PNCounter {

    private final Map<String, Long> increments;
    private final Map<String, Long> decrements;

    public PNCounter() {
        this.increments = new HashMap<>();
        this.decrements = new HashMap<>();
    }

    private PNCounter(Map<String, Long> increments, Map<String, Long> decrements) {
        this.increments = new HashMap<>(increments);
        this.decrements = new HashMap<>(decrements);
    }

    public void increment(String nodeId) {
        increments.merge(nodeId, 1L, Long::sum);
    }

    public void decrement(String nodeId) {
        decrements.merge(nodeId, 1L, Long::sum);
    }


    public long value() {
        long inc = increments.values().stream().mapToLong(Long::longValue).sum();
        long dec = decrements.values().stream().mapToLong(Long::longValue).sum();
        return inc - dec;
    }


    public void merge(PNCounter other) {
        for (var e : other.increments.entrySet()) {
            increments.merge(e.getKey(), e.getValue(), Math::max);
        }
        for (var e : other.decrements.entrySet()) {
            decrements.merge(e.getKey(), e.getValue(), Math::max);
        }
    }

    // für Tests
    public PNCounter copy() {
        return new PNCounter(increments, decrements);
    }

    public Map<String, Long> increments() {
        return Map.copyOf(increments);
    }

    public Map<String, Long> decrements() {
        return Map.copyOf(decrements);
    }
}
